### ISSUE TITLE
Support async function

### ISSUE THREAD

**Issue by SeyyedKomeil** [celery/celery#6552]
# Checklist

- [x] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Issue+Type%3A+Feature+Request%22+)
  for similar or identical feature requests.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3A%22PR+Type%3A+Feature%22+)
  for existing proposed implementations of this feature.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the if the same feature was already implemented in the
  master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

# Brief Summary

Support async celery tasks & call async function in celery tasks

# Design
```
async def foo(x, y):
    return x + y

@app.task(
    name='tasks.add',
    queue='custom',
    rate_limit='10/s',
)
async def add(x, y):
    await foo(x, y)
```

**Comment by auvipy** [comment]
its a big undertake

**Comment by SeyyedKomeil** [comment]
> its a big undertake

So any idea to call async function in celery task?
I use asyncio.run but sometimes got thread error.
@auvipy

**Comment by auvipy** [comment]
it's not currently supported but we are working on implementing it as soon as possible

**Comment by tasn** [comment]
Would be a dream. A lot of my tasks are IO bound rather than CPU bound. I was looking at maybe using eventlet, but it's not entirely obvious to me how it'll interact with asyncio (aiohttp and aiopg) which I'm using.

**Comment by auvipy** [comment]
you can read celery CEP on this. though I need to write a CEP with some POC

**Comment by Kludex** [comment]
> you can read celery CEP on this. though I need to write a CEP with some POC

Where is the CEP?

**Comment by auvipy** [comment]
we have to write one but the high level ones are here https://github.com/celery/ceps

**Comment by Kludex** [comment]
Is this open for work, or you'll personally create the CEP and the mentioned POC at some point? (not saying that I want to do it or will, just asking)

**Comment by u23a** [comment]
Will it appear in 6.0.0 Milestone?

**Comment by GuptaAman08** [comment]
Any rough estimation of when celery will support this going forward ??

**Comment by aleGpereira** [comment]
This will be nice to have, for sure.

**Comment by auvipy** [comment]
> Will it appear in 6.0.0 Milestone?

celery 6 will mainly focus on asyncio support. but as you all can see you can't do that before Q1 2023

**Comment by lucasgadams** [comment]
Are we still looking at Q3? Some semblance of asyncio support is the deciding factor for whether we can use celery or not. It is very badly needed for modern python

**Comment by bengabp** [comment]
Its been 3 years, is this now supported ?

**Comment by marcyvillegas** [comment]
The workaround I used related to this issue is [bblanchon](https://github.com/bblanchon)'s [reply ](https://github.com/celery/celery/issues/6603#issuecomment-817159091) from #6603 where he created a decorator which converts the task to sync. With that, you can `await` async functions inside a task.

**Comment by marcyvillegas** [comment]
The workaround I used related to this issue is Fran's [stackoverflow answer](https://stackoverflow.com/questions/72529998/how-to-use-async-function-inside-celery-task?answertab=scoredesc#tab-top).

His workaround:
1. Define decorator which converts async task to sync.
```py
from functools import wraps
from celery import Celery, Task
from typing import Any, Callable, Coroutine, ParamSpec, TypeVar
from asgiref import sync


_P = ParamSpec("_P")
_R = TypeVar("_R")


def async_task(app: Celery, *args: Any, **kwargs: Any):
    def _decorator(func: Callable[_P, Coroutine[Any, Any, _R]]) -> Task:
        sync_call = sync.AsyncToSync(func)

        @app.task(*args, **kwargs)
        @wraps(func)
        def _decorated(*args: _P.args, **kwargs: _P.kwargs) -> _R:
            return sync_call(*args, **kwargs)

        return _decorated

    return _decorator
```

2. Apply it on task
```py
import asyncio
import celery
from celery.utils.log import get_task_logger

from ..celery import app
from .async_task import async_task


@async_task(app, bind=True)
async def test_task(self: celery.Task):
    await asyncio.sleep(1.0)
    return "world"


@app.task(ignore_result=True)
def say_hello(who: str):
    print(f"Hello {who}")
```

**Comment by ehiggs** [comment]
@marcyvillegas I think that just converts the async task to a task in a sync task in a thread. This might mean you are able to use `async` tasks, but it doesn't get you the intended benefits of using async (lower overhead of concurrency than threads).

**Comment by onlinejudge95** [comment]
Any plans for supporting the same, I have a I/O bound task that would benefit a lot from using the same

**Comment by npip99** [comment]
There is a lot of GPT-4 use these days, and Celery can't handle this use case currently.

A potential workaround is to set Celery's max CPU cores to far beyond the actual number of CPU cores, i.e. utilize OS-level task management.

However by using that workaround, if you have GPT-4 tasks _and_ CPU tasks, a Celery instance can die if too many CPU tasks are taken on by a single machine.

The only alternative is to dedicate an entire CPU core to each task that's waiting on OpenAI for 30 seconds, which is very wasteful.

In essence, having an asyncio pool would be crucial for any Celery use case that depends on tasks completed by another machine, such as OpenAI calls. E.g. Consider a Celery implementation where you can do `def task()` to reserve a CPU core, and `async def task()` to join the asyncio pool, and the problem would be resolved.
- An idle core can decide to take on a sync task _or_ own an asyncio pool, and then go back to idle when it runs out of tasks. With this method, every CPU core is always doing something.

**Comment by ridiaz** [comment]
gevent pool works great with Celery with native support https://www.gevent.org/install.html .  You could run thousands of workers in a single machine if your tasks are IO-bound.  Monkey patching the std library seems like a hack initially, but it works like magic.

**Comment by lucasgadams** [comment]
@ridiaz if we have a code base that is written with asyncio, and the only thing that is not asyncio compatible is celery, is it possible to use gevent as a bridge to run our async functions in celery tasks? From my understanding (which isnt great on gevent, eventlet, ect) we are talking about patching non-async functions (like requests.get) rather than running already async compatible functions (aiohttp.get). From what I read SQLAlchemy was able to add asyncio support by using the greenlet library (https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html) and they said it wasn't that difficult. Could a similar approach be used here?

**Comment by chenyuanrun** [comment]
Need this feature too.

**Comment by jefer94** [comment]
I think that it should be easily covered, python has some functions that allow us to know if a function is a coroutine or an async generator, also, you should accept some argument in the app to see which lib (Asyncio/Trio/Uvloop) use and use by default Asyncio that provides `asyncio.run` to execute and wait and async code within a sync context.

It should use:
- asyncio.iscoroutinefunction(fn)
- inspect.isasyncgenfunction(fn)

If some of these libs don't include something like `asyncio.iscoroutinefunction` we should open an issue in that library to ask them for that feature.

For it and Python 3.13 it should start to use coroutines over threads, I think that the Python Coroutines aren't like Goroutines that use automatically all CPU threads, I don't know if this problem will be fixed in Python 3.13 but I think that it doesn't, generally, Python must manage it for us

**Comment by nicoeiris11** [comment]
Any updates on this feature? When should we expect async func exec support for our celery tasks?

**Comment by Nusnus** [comment]
@brylie-wolt 
> We're evaluating Celery alternatives and migrating away from Celery due to a lack of async support. It would be helpful in the evaluation process to know if there are any concrete plans or timelines related to Celery async support so I can include that information in internal discussions.
> 
> Is there at least a Celery Enhancement Proposal related to async support for review or some other signal of pending async support?

I'm sad to hear, but I can also understand.

Don't take this as an official announcement, but I've been investigating the native asyncio support gap recently to comprehend what we are even dealing with. I want to understand what is the scope of the effort, possible milestones and push forward to implementing the support (myself inc.) to release Celery v6.0.0 with asyncio pool.

The biggest hurdle is funding.
I have the time, skills and as a main maintainer I'm familiar with many of the internals already.

Does it make sense to raise a campaign to fund the development of this feature?
Or is there anyone from this discussion that is interested in sponsoring the work?

Let me know 🙏
tomer.nosrati@gmail.com

**Comment by brylie-wolt** [comment]
Thanks for the transparency @Nusnus. Setting up a funding campaign for the work is an excellent option, such as using GitHub Sponsors and promoting that campaign to get the job done. 

Thanks also for your effort in maintaining this crucial open-source project!

**Comment by dalberto** [comment]
I agree, setting up a campaign is worthwhile to get the appropriate funding. I've been following this issue for a couple of years, and I'd much rather support Celery than yet-another startup trying to deliver a half-baked solution.

**Comment by aleGpereira** [comment]
Celery is a wide-used library. I'm pretty sure many companies will be interested on making the async feature a reality.

**Comment by Nusnus** [comment]
Hey everyone, again, this is not an official announcement, but I made progress on my research, and as long as I am around this topic, I’ll use this issue to update/consult with the community.

I am highly considering starting with Kombu. This means making a `kombu.async` package with all of the new source, which will mimic and practically duplicate the standard synced API[1]. Theoretically, the interface should be the same, except using `async def` and `await` with the new `kombu.async` API.

Only once we have an async native Kombu (v6.0.0) that we can rely on, adding the matching Celery support makes sense to me.

It’s very brief, I know, but does it sound like the right direction?

[1] Avoiding code duplications as much as possible by only re-implementing IO-related code.

**Comment by Nusnus** [comment]
> Hey everyone, again, this is not an official announcement, but I made progress on my research, and as long as I am around this topic, I’ll use this issue to update/consult with the community.

👆

> I am highly considering starting with Kombu. This means making a `kombu.asyncio` package with all of the new source, which will mimic and practically duplicate the standard synced API[1]. Theoretically, the interface should be the same, except using `async def` and `await` with the new `kombu.asyncio` API.

I’ve been playing around, checking out my assumptions that I want to double-check with the community.
I’ve migrated the AMQP transport component to native asyncio and created a comparison benchmarking test.

1. `kombu.asyncio.transport.pyamqp` maintain the same interface as `kombu.transport.pyamqp`
2. Using `aio-pika` instead of `py-amqp`.
3. Similar example flows using sleep as an artificial network delay.

## Few questions
1. What’s your intuition says about this POC?
2. Does my benchmark logic make sense to you?
3. Can someone offer more benchmark functions for different edge cases? - You may only use `pyamqp.py` API and external libriaries for simulating environment issues.

## Sync (Standard) Example
```python
from __future__ import annotations

import time

from kombu.transport.pyamqp import Channel, Connection


def benchmark_sync() -> None:
    print("Running benchmark_sync with artificial delay…”)

    connection_params = {
        "hostname": “localhost”,
        "port": 5672,
        "userid": “guest”,
        "password": “guest”,
        "virtual_host": “/“,
    }

    num_iterations = 10_000
    delay = 0.001  # Simulated I/O delay in seconds

    connection = Connection(**connection_params)
    connection.connect()
    print("Connection established.”)

    channel: Channel = connection.channel()
    print("Channel opened.”)

    queue_name = “benchmark_queue”
    channel.queue_declare(queue=queue_name, durable=False, auto_delete=False)
    print(f"Queue '{queue_name}' declared.”)

    channel.queue_purge(queue=queue_name)
    print(f"Queue '{queue_name}' purged.”)

    start_time = time.perf_counter()

    for i in range(num_iterations):
        message_body = f"Benchmark Message {i}”
        message = channel.prepare_message(
            body=message_body,
            content_type="text/plain”,
            content_encoding="utf-8”,
        )

        channel.basic_publish(
            msg=message,
            exchange=“”,
            routing_key=queue_name,
        )
        time.sleep(delay)

    end_time = time.perf_counter()

    total_time = end_time - start_time
    avg_time = total_time / num_iterations

    print(f"Total time for {num_iterations} iterations: {total_time:.4f} seconds”)
    print(f"Average time per iteration: {avg_time:.6f} seconds”)

    connection.close()
    print("Connection closed.”)


def main() -> None:
    for i in range(10):
        print(f"\nIteration {i + 1} of benchmark_sync:”)
        benchmark_sync()


if __name__ == “__main__”:
    main()
```
### Runtime: 3m 6.47s
```console
python "examples/asyncio/asyncio_pyamqp(sync).py”

Iteration 1 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 19.7650 seconds
Average time per iteration: 0.001977 seconds
Connection closed.

Iteration 2 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 20.8292 seconds
Average time per iteration: 0.002083 seconds
Connection closed.

Iteration 3 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 17.9568 seconds
Average time per iteration: 0.001796 seconds
Connection closed.

Iteration 4 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 12.2152 seconds
Average time per iteration: 0.001222 seconds
Connection closed.

Iteration 5 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 18.7046 seconds
Average time per iteration: 0.001870 seconds
Connection closed.

Iteration 6 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 19.2657 seconds
Average time per iteration: 0.001927 seconds
Connection closed.

Iteration 7 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 19.3753 seconds
Average time per iteration: 0.001938 seconds
Connection closed.

Iteration 8 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 18.6518 seconds
Average time per iteration: 0.001865 seconds
Connection closed.

Iteration 9 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 19.5053 seconds
Average time per iteration: 0.001951 seconds
Connection closed.

Iteration 10 of benchmark_sync:
Running benchmark_sync with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 19.4217 seconds
Average time per iteration: 0.001942 seconds
Connection closed.
```
## Native Asyncio Example
```python
from __future__ import annotations

import asyncio
import time

from kombu.asyncio.transport.pyamqp import Channel, Connection


async def benchmark_asyncio() -> None:
    print("Running benchmark_asyncio with artificial delay…”)

    connection_params = {
        "host": “localhost”,
        "port": 5672,
        "login": “guest”,
        "password": “guest”,
        "virtualhost": “/“,
    }

    num_iterations = 10_000
    delay = 0.001  # Simulated I/O delay in seconds

    connection = Connection(**connection_params)
    await connection.connect()
    print("Connection established.”)

    channel: Channel = await connection.channel()
    print("Channel opened.”)

    queue_name = “benchmark_queue”
    queue = await channel.declare_queue(queue_name, durable=False, auto_delete=False)
    print(f"Queue '{queue_name}' declared.”)

    await queue.purge()
    print(f"Queue '{queue_name}' purged.”)

    exchange = channel.default_exchange

    start_time = time.perf_counter()

    async def publish_message(i):
        message_body = f"Benchmark Message {i}”
        message = channel.prepare_message(
            body=message_body,
            content_type="text/plain”,
            content_encoding="utf-8”,
        )

        await exchange.publish(
            message,
            routing_key=queue_name,
        )
        await asyncio.sleep(delay)

    tasks = [publish_message(i) for i in range(num_iterations)]

    await asyncio.gather(*tasks)

    end_time = time.perf_counter()

    total_time = end_time - start_time
    avg_time = total_time / num_iterations

    print(f"Total time for {num_iterations} iterations: {total_time:.4f} seconds”)
    print(f"Average time per iteration: {avg_time:.6f} seconds”)

    await connection.close()
    print("Connection closed.”)


async def main() -> None:
    for i in range(10):
        print(f"\nIteration {i + 1} of benchmark_asyncio:”)
        await benchmark_asyncio()


if __name__ == “__main__”:
    asyncio.run(main())
```
### Runtime: 17.556s
```console
python "examples/asyncio/asyncio_pyamqp.py”

Iteration 1 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6564 seconds
Average time per iteration: 0.000166 seconds
Connection closed.

Iteration 2 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.7135 seconds
Average time per iteration: 0.000171 seconds
Connection closed.

Iteration 3 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6795 seconds
Average time per iteration: 0.000168 seconds
Connection closed.

Iteration 4 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6132 seconds
Average time per iteration: 0.000161 seconds
Connection closed.

Iteration 5 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6795 seconds
Average time per iteration: 0.000168 seconds
Connection closed.

Iteration 6 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.7088 seconds
Average time per iteration: 0.000171 seconds
Connection closed.

Iteration 7 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6958 seconds
Average time per iteration: 0.000170 seconds
Connection closed.

Iteration 8 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6821 seconds
Average time per iteration: 0.000168 seconds
Connection closed.

Iteration 9 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.7346 seconds
Average time per iteration: 0.000173 seconds
Connection closed.

Iteration 10 of benchmark_asyncio:
Running benchmark_asyncio with artificial delay…
Connection established.
Channel opened.
Queue 'benchmark_queue’ declared.
Queue 'benchmark_queue’ purged.
Total time for 10000 iterations: 1.6117 seconds
Average time per iteration: 0.000161 seconds
Connection closed.
```

**Comment by auvipy** [comment]
1. The PoC seems pragmatic to me.

2. The Benchmark seems legit to me

**Comment by aleGpereira** [comment]
Thank you for taking the time for this!

For now the result speak by itself. By switching to async, you made an improvement of 15x-20x times faster. That's really high.

> What’s your intuition says about this POC?

It's a good start.

> Does my benchmark logic make sense to you? 

It does to me. Unfortunately I'm not a amqp expert so is hard for me to suggest more effective tests.

> Can someone offer more benchmark functions for different edge cases?

I guess Celery development team has already gone through this when the sync library was developed. What about using that experience/history for this purpose? I guess there should be edge cases written somewhere?

**Comment by Nusnus** [comment]
> Thank you for taking the time for this!
> 
> For now the result speak by itself. By switching to async, you made an improvement of 15x-20x times faster. That's really high.
> 
> > What’s your intuition says about this POC?
> 
> It's a good start.
> 
> > Does my benchmark logic make sense to you? 
> 
> It does to me. Unfortunately I'm not a amqp expert so is hard for me to suggest more effective tests.
> 
> > Can someone offer more benchmark functions for different edge cases?
> 
> I guess Celery development team has already gone through this when the sync library was developed. What about using that experience/history for this purpose? I guess there should be edge cases written somewhere?

Thank you for the support 🙏
Unfortunately, the original code is about a decade old, so I'll have to deal with what we have. Np though.

**Comment by Nusnus** [comment]
> For now the result speak by itself. By switching to async, you made an improvement of 15x-20x times faster. That's really high.
> 

Tbh this really hypes me up. I'm eager to keep going and I'm actually having a lot of fun, learning a lot about the internals with my research and seeing amazing results indeed!

**Comment by ehiggs** [comment]
@Nusnus thanks for pushing on this! 

>Does my benchmark logic make sense to you?

In this benchmark you compare serially running sync tasks against the rabbit broker vs blasting all the async tasks against rabbit at once. But celery supports threaded execution. 

I *think* you want something more like the following (which I haven't tried/tested):

```python
def publish_message(i):
    message_body = f"Benchmark Message {i}”
    message = channel.prepare_message(
        body=message_body,
        content_type="text/plain”,
        content_encoding="utf-8”,
    )

    channel.basic_publish(
        msg=message,
        exchange=“”,
        routing_key=queue_name,
    )
    time.sleep(delay)

max_workers = 20 # select however many threads you want doing the work
with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
    # start timer here
    tasks = [executor.submit(publish_message, i) for i in range(num_iterations)]
    concurrent.futures.wait(tasks)
    # stop timer here.
```

**Comment by Nusnus** [comment]
@ehiggs 
> @Nusnus thanks for pushing on this!
> 
> > Does my benchmark logic make sense to you?
> 
> In this benchmark you compare serially running sync tasks against the rabbit broker vs blasting all the async tasks against rabbit at once. But celery supports threaded execution.
> 
> I _think_ you want something more like the following (which I haven't tried/tested):
> 
> ```python
> def publish_message(i):
>     message_body = f"Benchmark Message {i}”
>     message = channel.prepare_message(
>         body=message_body,
>         content_type="text/plain”,
>         content_encoding="utf-8”,
>     )
> 
>     channel.basic_publish(
>         msg=message,
>         exchange=“”,
>         routing_key=queue_name,
>     )
>     time.sleep(delay)
> 
> max_workers = 20 # select however many threads you want doing the work
> with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
>     # start timer here
>     tasks = [executor.submit(publish_message, i) for i in range(num_iterations)]
>     concurrent.futures.wait(tasks)
>     # stop timer here.
> ```
> 
> Edit: you may need to create the channel inside publish_message as channels may not be thread safe in pyamqp.

You’re definitely correct, but let me try to convey a point here, and tell me what you think.

I’ve created an optimized version of the sync code. The optimized runtime goes down from ~3m to 9.651s.
That being said, let’s review the code to see how much did we need to complicate the current API to reach the asyncio native results.

## Optimized Code
```python
from __future__ import annotations

import itertools
import math
import threading
import time

from kombu.transport.pyamqp import Channel, Connection


def benchmark_sync_optimized() -> None:
    print("Running optimized benchmark_sync...")

    connection_params = {
        "hostname": "localhost",
        "port": 5672,
        "userid": "guest",
        "password": "guest",
        "virtual_host": "/",
    }

    num_iterations = 10_000
    delay = 0.001  # Simulated I/O delay in seconds

    setup_connection = Connection(**connection_params)
    setup_connection.connect()
    print("Setup connection established.")

    queue_name = "benchmark_queue_sync_optimized"
    setup_channel: Channel = setup_connection.channel()
    print("Setup channel opened.")

    setup_channel.queue_declare(queue=queue_name, durable=False, auto_delete=False)
    print(f"Queue '{queue_name}' declared.")

    setup_channel.queue_purge(queue=queue_name)
    print(f"Queue '{queue_name}' purged.")

    setup_channel.close()
    print("Setup channel closed.")
    setup_connection.close()
    print("Setup connection closed.")

    max_workers = 100

    def publish_messages(thread_id, messages):
        try:
            connection = Connection(**connection_params)
            connection.connect()
            channel: Channel = connection.channel()
            # print(f"Thread {thread_id}: Connection and channel established.")

            for i in messages:
                message_body = f"Benchmark Message {i}"
                message = channel.prepare_message(
                    body=message_body,
                    content_type="text/plain",
                    content_encoding="utf-8",
                )
                channel.basic_publish(
                    msg=message,
                    exchange="",
                    routing_key=queue_name,
                )
                time.sleep(delay)

            channel.close()
            # print(f"Thread {thread_id}: Channel closed.")
            connection.close()
            # print(f"Thread {thread_id}: Connection closed.")
        except Exception as e:
            print(f"Error in publish_messages in thread {thread_id}: {e}")

    start_time = time.perf_counter()

    # Distribute messages evenly among threads
    messages_per_thread = math.ceil(num_iterations / max_workers)
    message_ranges = list(itertools.batched(range(num_iterations), messages_per_thread))

    threads = []
    for thread_id, messages in enumerate(message_ranges):
        t = threading.Thread(target=publish_messages, args=(thread_id, messages))
        threads.append(t)
        t.start()
        # print(f"Thread {thread_id}: Started.")

    # Wait for all threads to complete
    for t in threads:
        t.join()
        # print(f"Thread {threads.index(t)}: Completed.")

    end_time = time.perf_counter()
    total_time = end_time - start_time
    avg_time = total_time / num_iterations
    print(f"Total time for {num_iterations} iterations: {total_time:.4f} seconds")
    print(f"Average time per iteration: {avg_time:.6f} seconds")

    print("Benchmark complete.”)
```
### Runtime: 9.651s
```console
python "examples/asyncio/asyncio_pyamqp(sync).py"

Iteration 1 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.9622 seconds
Average time per iteration: 0.000096 seconds
Benchmark complete.

Iteration 2 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 1.0062 seconds
Average time per iteration: 0.000101 seconds
Benchmark complete.

Iteration 3 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 1.1053 seconds
Average time per iteration: 0.000111 seconds
Benchmark complete.

Iteration 4 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.8148 seconds
Average time per iteration: 0.000081 seconds
Benchmark complete.

Iteration 5 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.7966 seconds
Average time per iteration: 0.000080 seconds
Benchmark complete.

Iteration 6 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.8028 seconds
Average time per iteration: 0.000080 seconds
Benchmark complete.

Iteration 7 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.8451 seconds
Average time per iteration: 0.000085 seconds
Benchmark complete.

Iteration 8 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.7389 seconds
Average time per iteration: 0.000074 seconds
Benchmark complete.

Iteration 9 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.9046 seconds
Average time per iteration: 0.000090 seconds
Benchmark complete.

Iteration 10 of benchmark_sync_optimized:
Running optimized benchmark_sync...
Setup connection established.
Setup channel opened.
Queue 'benchmark_queue_sync_optimized' declared.
Queue 'benchmark_queue_sync_optimized' purged.
Setup channel closed.
Setup connection closed.
Total time for 10000 iterations: 0.8977 seconds
Average time per iteration: 0.000090 seconds
Benchmark complete.
```

Whereas the asyncio code is much simpler;
### Asyncio Example
```python
async def benchmark_asyncio() -> None:
    print("Running benchmark_asyncio with artificial delay...")

    connection_params = {
        "host": "localhost",
        "port": 5672,
        "login": "guest",
        "password": "guest",
        "virtualhost": "/",
    }

    num_iterations = 10_000
    delay = 0.001  # Simulated I/O delay in seconds

    connection = Connection(**connection_params)
    await connection.connect()
    print("Connection established.")

    channel: Channel = await connection.channel()
    print("Channel opened.")

    queue_name = "benchmark_queue_asyncio"
    queue = await channel.declare_queue(queue_name, durable=False, auto_delete=False)
    print(f"Queue '{queue_name}' declared.")

    await queue.purge()
    print(f"Queue '{queue_name}' purged.")

    exchange = channel.default_exchange

    start_time = time.perf_counter()

    async def publish_message(i):
        message_body = f"Benchmark Message {i}"
        message = channel.prepare_message(
            body=message_body,
            content_type="text/plain",
            content_encoding="utf-8",
        )

        await exchange.publish(
            message,
            routing_key=queue_name,
        )
        await asyncio.sleep(delay)

    tasks = [publish_message(i) for i in range(num_iterations)]

    await asyncio.gather(*tasks)

    end_time = time.perf_counter()

    total_time = end_time - start_time
    avg_time = total_time / num_iterations

    print(f"Total time for {num_iterations} iterations: {total_time:.4f} seconds")
    print(f"Average time per iteration: {avg_time:.6f} seconds")

    await connection.close()
    print("Connection closed.”)
```

Lastly, of course these tests are very specific, but IMHO they validate that we’re on the right path.

**Comment by Nusnus** [comment]
> @Nusnus thanks for pushing on this! 
> 

Thank you 🙏
Since I joined the Celery team a few years ago, I've been very active towards increasing the QA capabilities and crushing bugs, in addition to closing gaps and making sure everything flows with the organization. I've matured a lot in the Celery domain and had to dive deep so many times into the unknown.

I'm fascinated by the challenge of adding asyncio support and bringing Celery to the front lines, hopefully fixing and improving long due issues along the way as well.

I just hope I'll find a way to fund the effort before I exhaust my current (and very limited) resources 🙏

---

Anyone interested in sponsoring the work, please contact me at:
tomer.nosrati@gmail.com

**Comment by auvipy** [comment]
we can check how recently they approached this https://www.psycopg.org/articles/2024/09/23/async-to-sync/ . Also it is not all about the speed only. celery should have the native asyncio/anyio supported

**Comment by Nusnus** [comment]
> we can check how recently they approached this https://www.psycopg.org/articles/2024/09/23/async-to-sync/ . Also it is not all about the speed only. celery should have the native asyncio/anyio supported

Thank you, I’ll check it out!
And of course. Asyncio native means what it means.
The speed though is a huge incentive :)

**Comment by ehiggs** [comment]
>Also it is not all about the speed only. celery should have the native asyncio/anyio supported

I agree it's not all about the speed. I think it's underappreciated that thread and gevent models do not support [timeouts](https://docs.celeryq.dev/en/stable/userguide/workers.html#time-limits). And prefork only supports it by killing the process.

**Comment by Andrew-Chen-Wang** [comment]
Would recommend reaching out to elasticsearch-py team (sethmlarson) and see what they came up with re duplicating code and maintainability; their approach became a package: `unasync`, taking async code to sync code assuming you wouldn't need `asyncio.gather()`.

My main ask for Celery maintainers is to make an asyncio-compatible interface since the celery tasks themselves are where the performance bottlenecks are. Thanks for all your guys' time!

**Comment by Nusnus** [comment]
> Would recommend reaching out to elasticsearch-py team (sethmlarson) and see what they came up with re duplicating code and maintainability; their approach became a package: `unasync`, taking async code to sync code assuming you wouldn't need `asyncio.gather()`.
> 
> My main ask for Celery maintainers is to make an asyncio-compatible interface since the celery tasks themselves are where the performance bottlenecks are. Thanks for all your guys' time!

Great tip!
Thank you!!

Keep’em coming, it only motivates me more! 😁

**Comment by lucasgadams** [comment]
We decided we couldn't wait any longer for asyncio support in celery. I appreciate the recent movement to try and make some progress, but it just doesn't seem like this is a priority or going to be done in a reasonable timeline. We've migrated to Dramatiq which was able to pretty easily add asyncio support this year, and so far has been a good drop in replacement. Will keep following to see what happens in the future.

**Comment by xucian** [comment]
any new plans on this?

**Comment by Nusnus** [comment]
@xucian 
> any new plans on this?

Yes.

> Hey everyone, again, this is not an official announcement, but I made progress on my research, and as long as I am around this topic, I’ll use this issue to update/consult with the community.
> 
> I am highly considering starting with Kombu. This means making a `kombu.asyncio` package with all of the new source, which will mimic and practically duplicate the standard synced API[1]. Theoretically, the interface should be the same, except using `async def` and `await` with the new `kombu.asyncio` API.
> 
> Only once we have an asyncio native Kombu (v6.0.0) that we can rely on, adding the matching Celery support makes sense to me.
> 
> It’s very brief, I know, but does it sound like the right direction?
> 
> [1] Avoiding code duplications as much as possible by only re-implementing IO-related code.

I have actually finished my research, and mapped the MVP components of both Kombu and Celery, to migrate to AsyncIO.
My plan propose migrating the RabbitMQ broker and RPC backend for the first version, which will run alongside the sync API for the entire lifetime of v6.x.x, with potentially deprecating the sync API for v7.x.x/8.x.x, after all of the functionality is migrated. The goal is fully native asyncio support from the transport to the tasks themselves.

In addition, I have compiled a coherent work plan for both Kombu v6.0.0 and Celery v6.0.0 and defined a work methodology for the project that involves the community to allow input from active contributors and users. The plan propose an async development approach between Kombu and Celery, for smooth internals integration, which is optimized for a single feature owner and pragmatic milestones for each component.

My work plan is split to deliverables per milestone, migrating the layers one by one from the transport above, following an agile approach and an MVP scope limit.

I would be happy to engage with anyone willing to sponsor (**partially** as well) this project, and provide the full details of my plan.

I am available at,
- tomer.nosrati@gmail.com
- https://www.linkedin.com/in/tomernosrati
- https://x.com/smilingnosrati

Please contact me - I am passionate about this project and would love to bring Celery to the modern age.

Thank you 🙏

P.S
@auvipy FYI

**Comment by holvianssi** [comment]
On the original issue, where the issue pointed out was that if you mix CPU heavy and IO wait heavy tasks, you'll be in trouble with non-async celery.

I'm afraid that's going to be the case with asyncio, too! Because if you launch gazillion CPU heavy tasks with asyncio, each task will progress slow in parallel, instead of fast per task sequentially.

So, you'll need some logic to process small amount of CPU heavy tasks in parallel, but allow for large amount of IO heavy tasks in parallel, even if you are using asyncio. So, in. the end, you'll likely need separate queues even with asyncio support.

**Comment by davidemerritt** [comment]
In the meantime while this is being worked on, is there any known issue with using something like `AsyncToSync` from https://pypi.org/project/asgiref/  to run coroutine async code within a gthread type celery worker?

**Comment by psinojiya** [comment]
any new plans or update on this?

**Comment by giuliotosato** [comment]
updates?

**Comment by auvipy** [comment]
yes, we will have this supported very soon. most probably In v5.7. and I will handle this personally.

**Comment by IvoPereira89** [comment]
@auvipy Do you already have a sense o when this may become available?

**Comment by kunaldeo** [comment]
Let the man work on it. Why so much hurry on this?

**Comment by auvipy** [comment]
I will start on this after 5.6rc2 release, which is due this month.

**Comment by InfernalAzazel** [comment]
> I will start on this after 5.6rc2 release, which is due this month.

How is the current progress? We are very much looking forward to the asynchronous task being able to run

**Comment by NathanAP** [comment]
Sorry to revive, but its been 6 years of "as soon as possible". At this point I'm hopeless.

**Comment by alexted** [comment]
Dude, chill out, did you just fall from the moon? Technologies come and go - that's how it's always been, is now, and will be. 
Don't hyperfixate on this library, look around and you'll see how many interesting things there are.
Personally, I've found at least two alternatives for myself:
- [Taskiq](https://github.com/taskiq-python/taskiq)
- [FastWorker](https://github.com/neul-labs/fastworker)

**Comment by NathanAP** [comment]
Sorry if this hurt you somehow (not sure why or how), but its the truth. 

Celery is one of the biggest and most important open source libraries we have for Python. If you cannot deal with your "promises" then don't create expectations. This post is the one with the most thumbs up votes they have, it should be their priority, but for half a decade it still isn't.

**Comment by alexted** [comment]
I totally feel your frustration, and honestly, I’m right there with you. But over time, I’ve realized that life is just more complex than we’d like it to be. We aren’t robots, we’re human - imperfect, sometimes overwhelmed, and often irrational. Things rarely go exactly the way we feel they "should".
Progress never stops, and no matter how great the library is, there will always be a successor. Sooner or later, everything has its season. Nothing lasts forever.
It’s also helpful to remember that open source is a labor of love. At the end of the day, it’s all voluntary, and no one truly owes anyone anything.
If you feel like your expectations aren't being met, my best advice is to let go of that disappointment and move forward. It’s not worth the stress. As I mentioned before, the Python ecosystem is huge, and there are plenty of great alternatives out there for whatever you need!

And last but not least - don't forget that you can always help the community yourself and contribute to the project!

**Comment by ghirailghiro** [comment]
Opened celery/celery#10634 — an asyncio execution pool that runs coroutine
task bodies on one event loop shared by the worker process. It doesn't
touch the broker or the result backend, so I don't think it conflicts
with a native design for 6.0 — scope and the tradeoffs are in the PR
description.

cc @auvipy @Nusnus, since you've both discussed scope and design here
before — curious for your take.
